lib/executor/base_executor.go

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package executor
+
+import (
+	"context"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/ui/pb"
```

Comment:
Minor: currently `github.com/loadimpact/k6/ui/pb` is a dependency in a lot of places:

```sh
> rg -g '!*_test.go' -g '!vendor' 'k6/ui/pb"' -l
cmd/run.go
cmd/ui.go
cmd/cloud.go
lib/executors.go
lib/executor/variable_arrival_rate.go
lib/executor/per_vu_iterations.go
lib/executor/base_executor.go
lib/executor/helpers.go
lib/executor/variable_looping_vus.go
lib/executor/constant_arrival_rate.go
lib/executor/shared_iterations.go
lib/executor/constant_looping_vus.go
lib/executor/externally_controlled.go
core/local/local.go
```

It would be cleaner if UI handling is done only in `cmd/ui.go` (or preferably in a `cmd/ui` package).

Perhaps all executors could receive an `io.Writer` or even better: an out channel to send progress updates over. This would allow finer grained control over the output (buffering, aggregation, etc.), simplify and shorten the `Run` method for all executors, but also make these methods more testable, and separate the logic vs. presentation concerns better.

Reply:
> It would be cleaner if UI handling is done only in cmd/ui.go (or preferably in a cmd/ui package).

Wait, but why? :smile: In my mind, the more things we can remove from `cmd/`, the better. The simple rationale behind this statement is that the `cmd/` files are the entrypoints of the program. Execution starts from them and flows outward. They import everything and, because import loops are sensibly forbidden in Go, nothing can import them. They are the ones that deal with global state - way too much of it, at the moment, but that's another topic... But because of all of these reasons, the farther a piece of code is from `cmd/`, the more testable and self-contained it actually is.

You've mentioned how many things import the progress bar code (`ui/pb`). But why is that a problem in any way? To me, the reverse is much more important - how many things does `ui/pb` import? Because the less things it imports, the more self-contained it is, and thus - the easier it is to test, and the better it serves as a building block of other things. And with the current architecture, keeping in mind that there's still some work to be done on the progress bars, they have absolutely no dependence on other parts of k6! Running `go list -f '{{ join .Imports "\n" }}'  github.com/loadimpact/k6/ui/pb | sort` produces:
```
fmt
github.com/loadimpact/k6/vendor/github.com/fatih/color
math
strconv
strings
sync
time
```

So, from that perspective, the current `pb` is perfect as a building block - it can be used from every other part of k6.

To address your other concerns:
> Perhaps all executors could receive an `io.Writer` or even better: an out channel to send progress updates over. 

A key architectural decision in the current way of measuring executor progress is that it's a pull model. The executors don't push anything - pushing every single iteration/VU change/elapsed decisecond over some channel, or outputting it to some `io.Writer`, would be super inefficient! And it would also be a nightmare to synchronize for whatever is dealing with the multiple executors.

> This would allow finer grained control over the output (buffering, aggregation, etc.), simplify and shorten the `Run` method for all executors, but also make these methods more testable, and separate the logic vs. presentation concerns better.

The current executors IMO (though I'm biased :sweat_smile: ) offer a much simpler, more efficient, and easier to synchronize mechanism - simple callbacks. This shifts everything from a push-based model to a pull-based one, so there's nothing to buffer or aggregate. Even better, when you don't need the data, you don't needlessly receive or process it! So, by default, instead of processing every single executor event in some way, we only query the executors for their progress every [100ms](https://github.com/loadimpact/k6/blob/3421078d34e9eb2dce507c2687eeb3981d1afc87/cmd/ui.go#L127). But that's just when we're in the very interactive k6 execution mode, with the nice UI that should be smooth. If we're running without a TTY (e.g. in a CI setting), we query the executors [once every second](https://github.com/loadimpact/k6/blob/3421078d34e9eb2dce507c2687eeb3981d1afc87/cmd/ui.go#L123) - 10 times less often! And when we've run k6 [with `--quiet`](https://github.com/loadimpact/k6/blob/3421078d34e9eb2dce507c2687eeb3981d1afc87/cmd/ui.go#L107-L109), we don't query at all!

And true, there's some LoC cost for having it in the `Run()` method, but as shown by [this commit](https://github.com/loadimpact/k6/pull/1007/commits/2ebe3dfdd6c3ef97decbd9ac80c6747dedb3aac1#diff-c9c33569514bbb8c48eefc919f3cc32cR512), the `Run()` method can be refactored so that the progress tracking [isn't a direct part of it](https://github.com/loadimpact/k6/blob/3421078d34e9eb2dce507c2687eeb3981d1afc87/lib/executor/externally_controlled.go#L396-L410). And while in hindsight I fully agree that it was worth to do it for the `externally-controlled` executor in that commit, I'm not sure if it's worth to do for the simpler executors. For them, since they're much simpler, I'm on the fence if the complexity of the needed [additional executor state struct](https://github.com/loadimpact/k6/blob/3421078d34e9eb2dce507c2687eeb3981d1afc87/lib/executor/externally_controlled.go#L364) would be offset by the slightly improved readability and conciseness of their `Run()` methods...

> It would be cleaner if UI handling is done only in `cmd/ui.go` (or preferably in a `cmd/ui` package).

While still strongly disagreeing with the `cmd` part of that (see the start of this wall of text :sweat_smile: ), I somewhat agree that the executors shouldn't be very concerned with "UI", i.e. the concepts of presentation and progress bars. And, to be fair, they are currently not - they are only concerned with having some way to expose their internal progress in terms of a completion percentage and human-readable text, which to me seems appropriate, since these are inherent properties of running executors. 

The only rub comes from the fact that [this function](https://github.com/loadimpact/k6/blob/new-schedulers/ui/pb/progressbar.go#L42) currently sits in a package with the `ui` name. And while I can refactor the code and separate the pure progress reporting (which executors need and use) from the UI parts (which they currently don't use), I'm not sure this would be an improvement in terms of readability. I might be wrong, but it seems to me like it would be just more code that doesn't accomplish anything significant other than improving the ideological purity of the codebase, for its own sake :sweat_smile: .

Reply:
> the more things we can remove from `cmd/`, the better

True, definitely agree with that (there should be a generic `k6/ui` package), but considering `cmd/ui.go` was introduced here, what I meant was that it would be cleaner if all progress bar handling is also done there, instead of each executor importing `ui/pb`.

But you're right that `ui/pb` is self-contained, and the `progresFn`s defined in each executor are fairly generic and could be used for non-CLI UIs. That was my main concern, that we'd be coupling executors to UI elements, but `k6/ui` is not tied to the CLI, which is great.

I didn't consider the push vs. pull efficiencies, but I see your point. And actually quite like the `progresFn`/`snapshot` approach. :)

Thanks for explaining!